File: core/src/com/qtech/bubbleblaster/common/ability/AbilityType.java
package com.qtech.bubbleblaster.common.ability;

import com.qtech.bubbleblaster.common.RegistryEntry;

import java.util.Objects;
import java.util.function.Supplier;

/**
 * @param <T>
 */
public class AbilityType<T extends Ability<T>> extends RegistryEntry {
    // Types
//    public static final AbilityType<TeleportAbility> TELEPORT_ABILITY = new AbilityType<>(TeleportAbility.class, "teleport_ability");

    // Fields
    private final Supplier<T> ability;

    public AbilityType(Supplier<T> ability) {
        this.ability = ability;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AbilityType<?> that = (AbilityType<?>) o;
        return Objects.equals(getRegistryName(), that.getRegistryName());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getRegistryName());
    }

    public T getAbility() {
        return ability.get();
    }
}

File: src/main/kotlin/qtech/bubbles/common/ability/AbilityType.kt
package qtech.bubbles.common.ability

import qtech.bubbles.common.RegistryEntry
import java.util.*
import java.util.function.Supplier

/**
 * @param <T>
</T> */
class AbilityType<T : Ability<*>?>(// Types
    //    public static final AbilityType<TeleportAbility> TELEPORT_ABILITY = new AbilityType<>(TeleportAbility.class, "teleport_ability");
    // Fields
    private val ability: Supplier<T>
) : RegistryEntry() {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AbilityType<*>
        return registryName == that.registryName
    }

    override fun hashCode(): Int {
        return Objects.hash(registryName)
    }

    fun getAbility(): T {
        return ability.get()
    }
}